builder.go

```diff
@@ -75,110 +90,186 @@ func (b *Builder) Build() (*BuildMetadata, error) {
 	}
 	defer os.RemoveAll(planDir)
 
-	procMap := processMap{}
-	plan := b.Plan
-	var bom []BOMEntry
-	var slices []layers.Slice
-	var labels []Label
+	useSnapshotter := b.Snapshotter != nil
 
-	for _, bp := range b.Group.Group {
-		bpInfo, err := bp.Lookup(b.BuildpacksDir)
-		if err != nil {
+	if useSnapshotter {
+		if err := b.Snapshotter.Init(); err != nil {
 			return nil, err
 		}
-		bpDirName := launch.EscapeID(bp.ID)
-		bpLayersDir := filepath.Join(layersDir, bpDirName)
-		bpPlanDir := filepath.Join(planDir, bpDirName)
-		if err := os.MkdirAll(bpLayersDir, 0777); err != nil {
-			return nil, err
+	}
+
+	for _, bp := range b.Group.Group {
+		if useSnapshotter {
+			// first, we apply the existing snapshots - if they exist from previous cache
+			// this allows snapshot buildpacks to have cache and opt to do nothing
+			err := bp.ApplyLayerSnapshots(b.Snapshotter, b.LayersDir)
+			if err != nil {
+				return nil, err
+			}
 		}
 
-		if err := os.MkdirAll(bpPlanDir, 0777); err != nil {
+		launchData, newPlan, bpBOM, err := b.build(bp, appDir, layersDir, platformDir, planDir, plan)
+		if err != nil {
 			return nil, err
 		}
-		bpPlanPath := filepath.Join(bpPlanDir, "plan.toml")
 
-		foundPlan := plan.find(bp.noAPI())
-		if api.MustParse(bp.API).Equal(api.MustParse("0.2")) {
-			for i := range foundPlan.Entries {
-				foundPlan.Entries[i].convertMetadataToVersion()
+		if useSnapshotter {
+			bpDirName := launch.EscapeID(bp.ID)
+			bpLayersDir := filepath.Join(b.LayersDir, bpDirName)
+			layerDir := filepath.Join(bpLayersDir, "snapshot")
+			if err := os.MkdirAll(layerDir, 0777); err != nil {
+				return nil, err
 			}
-		}
-		if err := WriteTOML(bpPlanPath, foundPlan); err != nil {
-			return nil, err
-		}
 
-		cmd := exec.Command(
-			filepath.Join(bpInfo.Path, "bin", "build"),
-			bpLayersDir,
-			platformDir,
-			bpPlanPath,
-		)
-		cmd.Dir = appDir
-		cmd.Stdout = b.Out.Writer()
-		cmd.Stderr = b.Err.Writer()
-
-		if bpInfo.Buildpack.ClearEnv {
-			cmd.Env = b.Env.List()
-		} else {
-			cmd.Env, err = b.Env.WithPlatform(platformDir)
-			if err != nil {
+			snapshotPath := filepath.Join(layerDir, "snapshot.tgz")
+			if err := b.Snapshotter.TakeSnapshot(snapshotPath); err != nil {
 				return nil, err
 			}
+			layerTOMLPath := filepath.Join(bpLayersDir, "snapshot.toml")
```

Comment:
I noticed `snapshot.toml` isn't captured in the RFC, would that be worth adding?

Reply:
`snapshot.toml` is actually just a `layer.toml` for the generated `snapshot` layer. 🤯